### PR TITLE
BUG: Int Auto Conversion

### LINKED ISSUE #319: BUG: df.values[0] doesn't return an integer type
### Describe the bug
`df.values[0]` doesn't return an integer type

### To Reproduce
To help us to reproduce this bug, please provide information below:

Python: 3.8.8 (tags/v3.8.8:024d805, Feb 19 2021, 13:18:16) [MSC v.1928 64 bit (AMD64)] on win32
xorbits: 0.2.1

Python code:
```import xorbits
xorbits.init()
from xorbits import pandas as pd
t=pd.DataFrame({"ttt":range(10,20)})
m=t["ttt"].head(1).values[0]
for i in range(m):
    print(i)
```

Stack:
```
      3 t=pd.DataFrame({"ttt":range(10,20)})
      4 m=t["ttt"].head(1).values[0]
----> 5 for i in range(m):
      6     print(i)

TypeError: 'DataRef' object cannot be interpreted as an integer
```

### PR DESCRIPTION
**Pull request by Zhou1213CN**
## Related issue number

Fixes #319 

## Check code requirements

- [x] tests added / passed (if needed)
- [x] Ensure all linting tests pass

### CODE REVIEW

File: python/xorbits/core/data.py [hunk 1]
```diff
@@ -253,6 +253,33 @@ def __repr__(self):
...
+        from .execution import run
+
+        if (
+            self.data.data_type == DataType.tensor
+            and self.data._mars_entity.shape == ()
+            and self.data._mars_entity.dtype == "int"
+        ):
+            run(self)
+            return self.to_numpy()
+        elif self.data.data_type == DataType.object_ and self.to_object().isnumeric():
```
Comment (aresnow1) [inline python/xorbits/core/data.py:266]: `self.to_object()` will trigger execution, we don't need to check if it is numeric.

File: python/xorbits/core/data.py [hunk 2]
```diff
@@ -253,6 +253,33 @@ def __repr__(self):
...
+        else:
+            data = (
+                self.__str__()
+                if self.data.data_type != DataType.object_
+                else self.to_object().__str__()
+            )
+            raise TypeError(
+                "int() argument must be a string, a bytes-like object or a real number, not "
+                + data
+            )
```
Comment (aresnow1) [inline python/xorbits/core/data.py:278]:
```suggestion
            raise TypeError(
                f"int() argument must be a string, a bytes-like object or a real number, not {self.data.data_type}."
            )
```

File: python/xorbits/core/data.py [hunk 3]
```diff
@@ -253,6 +253,33 @@ def __repr__(self):
...
+
+    def __index__(self):
```
Comment (aresnow1) [inline python/xorbits/core/data.py:280]: It can be moved to next PR.

File: python/xorbits/core/data.py [hunk 4]
```diff
@@ -253,6 +253,33 @@ def __repr__(self):
...
+            run(self)
+            return self.to_numpy()
+        elif self.data.data_type == DataType.object_ and self.to_object().isnumeric():
+            run(self)
+            return int(self.to_object())
+        else:
+            data = (
+                self.__str__()
+                if self.data.data_type != DataType.object_
+                else self.to_object().__str__()
```
Comment (qianduoduo0904) [inline python/xorbits/core/data.py:273]: No `to_object()` which will trigger execution.

**Review by qianduoduo0904 — COMMENTED** [review]
Maybe the logic could be:

```
if hasattr(self, 'shape') and len(self.shape) == 0:
    # tensor type that is a scalar
    # ...
elif self.data_type == DataType.object:
    # run for object directly
else:
    raise error
```

File: python/xorbits/core/data.py [hunk 5]
```diff
@@ -253,6 +255,48 @@ def __repr__(self):
             run(self)
             return self.data.__repr__()
 
+    def _is_convertible_tensor(self):
+        if len(self.shape) == 0 and is_integer_dtype(self.dtype):
```
Comment (qianduoduo0904) [inline python/xorbits/core/data.py:259]:
```suggestion
    def _is_int_scalar_tensor(self):
        if self.data.data_type == DataType.tensor and len(self.shape) == 0 and is_integer_dtype(self.dtype):
```

File: python/xorbits/core/data.py [hunk 6]
```diff
@@ -253,6 +255,48 @@ def __repr__(self):
...
+            if isinstance(data_object, int):
+                return int(data_object)
+            else:
+                raise TypeError(
+                    f"{self.data.data_type} object cannot be interpreted as an integer."
+                )
+        else:
+            raise TypeError(
+                f"{self.data.data_type} object cannot be interpreted as an integer."
+            )
```
Comment (aresnow1) [inline python/xorbits/core/data.py:298]:
```suggestion
    def _to_int(self, cast=False):
        from .execution import run

        data_type = self.data.data_type
        if data_type == DataType.tensor and len(self.shape) == 0 and is_integer_dtype(self.dtype):
            run(self)
            return self.to_numpy()
        elif data_type == DataType.object_:
            run(self)
            data_object = self.to_object()
            if cast:
                return int(data_object)
            else:
                return data_object
        else:
            raise TypeError(
                f"{self.data.data_type} object cannot be interpreted as an integer."
            )
            
    def __int__(self):
        return self._to_int(cast=True)

    def __index__(self):
        try:
            val = self._to_int(cast=False)
            if isinstance(data_object, int):
                return val
            else:
                raise TypeError(
                    f"{self.data.data_type} object cannot be interpreted as an integer."
                )
        except TypeError:
            raise TypeError(
                f"{self.data.data_type} object cannot be interpreted as an integer."
            )
        
        
```

File: python/xorbits/core/data.py [hunk 7]
```diff
@@ -253,6 +255,48 @@ def __repr__(self):
...
+    def _is_convertible_tensor(self):
+        if len(self.shape) == 0 and is_integer_dtype(self.dtype):
+            return True
+        return False
+
+    def __int__(self):
+        from .execution import run
+
+        data_type = self.data.data_type
+        if data_type == DataType.tensor and self._is_convertible_tensor():
```
Comment (qianduoduo0904) [inline python/xorbits/core/data.py:267]: The check of data type can be put into function like I suggested above

File: python/xorbits/core/data.py [hunk 8]
```diff
@@ -253,6 +255,48 @@ def __repr__(self):
...
+        else:
+            raise TypeError(
+                f"int() argument must be a string, a bytes-like object or a real number, not {self.data.data_type}."
+            )
+
+    def __index__(self):
+        from .execution import run
+
+        data_type = self.data.data_type
+        if data_type == DataType.tensor and self._is_convertible_tensor():
```
Comment (qianduoduo0904) [inline python/xorbits/core/data.py:283]: Ditto.

File: python/xorbits/core/tests/test_execution.py [hunk 9]
```diff
@@ -119,6 +119,27 @@ def test_manual_execution(setup, dummy_int_series):
     assert not any([need_to_execute(ref) for ref in series_to_execute])
 
 
+def test_conversion(setup, dummy_int_2d_array, dummy_str_series):
```
Comment (qianduoduo0904) [inline python/xorbits/core/tests/test_execution.py:122]: You can reorder the tests to split them into parts like:

```
# int conversion
# tests about int()

# range
# tests about range()
```

File: python/xorbits/core/tests/test_execution.py [hunk 10]
```diff
@@ -119,12 +119,21 @@ def test_manual_execution(setup, dummy_int_series):
     assert not any([need_to_execute(ref) for ref in series_to_execute])
 
 
-def test_conversion(setup, dummy_int_2d_array, dummy_str_series):
+def tests_about_int(setup, dummy_int_2d_array):
```
Comment (qianduoduo0904) [inline python/xorbits/core/tests/test_execution.py:122]:
```suggestion
def tests_int_conversion(setup, dummy_int_2d_array):
```

File: python/xorbits/core/tests/test_execution.py [hunk 11]
```diff
@@ -119,12 +119,21 @@ def test_manual_execution(setup, dummy_int_series):
...
+    import xorbits.numpy as np
+    import xorbits.remote as xr
+
+    assert int(dummy_int_2d_array[0][2]) == 2
+    assert int(xr.spawn(lambda: "2")) == 2
+    with pytest.raises(TypeError):
+        int(np.atleast_2d(3.0))
+
+
+def tests_about_range(setup, dummy_int_2d_array, dummy_str_series):
```
Comment (qianduoduo0904) [inline python/xorbits/core/tests/test_execution.py:132]:
```suggestion
def tests_index_conversion(setup, dummy_int_2d_array, dummy_str_series):
```

**Review by aresnow1 — APPROVED** [review]
LGTM

**Review by qianduoduo0904 — APPROVED** [review]
LGTM

**Comment by qianduoduo0904** [comment]
Congrats, look forward to seeing more contributions from you.

**Pull request merged**